text/3446-store.md

```diff
@@ -0,0 +1,998 @@
+reasonably be implemented for `&dyn Store`, or `Rc<dyn Store>`, such is not the case for `Default`.
+
+This leaves 4 possibilities:
+
+-   Use `Clone` despite the poor semantics match.
+-   Use `Default` despite it being at odds with `dyn Store` use.
+-   Add a new `SpawningStore` trait to create an independent instance, though mixing several non-empty traits in a `dyn`
+    context is not supported yet.
+-   Add a method to `Store` to create an independent instance, fixing the semantics of `Clone`. Possibly a faillible
+    one.
+
+Note that technically switching from the `Clone` bound to another bound for `Box` and `Vec` is a breaking change,
+however since `Allocator` is an unstable API it is still early enough to effect such change.
+
+
+##  (Minor) To `Clone` or to share?
+
+As mentioned above, whenever an `Allocator` also implements the `Clone` trait, the clone or copy of the `Allocator` must
+fulfill specific requirements. In particular, all clones or copies of a given allocator behave as a single allocator
+sharing the backing memory and metadata. While the `Clone` trait does fit _creating_ a new clone/copy of an allocator,
+it is insufficient however to _query_ whether another instance is a clone/copy of a given allocator.
+
+The standard library runs headlong into this insufficience, and while `LinkedList::split_off` is implemented for any
+`Allocator` which also implements `Clone`, `LinkedList::append` is only implemented for `Global`.
+
+There are at least 2 possibilities, here:
+
+-   Add a requirement on `PartialEq` implementation for `Allocator` and `Store` that comparing equal means that they are
+    clones or copies of each others.
+-   Add a separate `SharingStore` trait -- see future possibilities.
+
+It should be noted that `dyn` usage of `Allocator` and `Store` suffers from the requirement of using unrelated traits as
+it is not possible to have a `dyn Allocator + Clone + PartialEq` trait today, though those traits can be implemented for
+`&dyn Allocator` or `Rc<dyn Allocator>`.
+
+Given that the problem is unsolved for `Allocator`, it can be punted on in the context of this RFC.
+
+
+##  (Minor) What should the capabilities of `Handle` be?
+
+Since any capability specified in the associated type definition is "mandatory", I am of the opinion that it should be
+kept to a minimum, and users can always specify additional requirements if they need to:
+
+-   At the moment, the only required is `Copy`. It could be downgraded to `Clone`, or removed entirely.
+    -   Although, do mind that just using `Store::grow`, or `Store::shrink` requires a copy/clone.
+-   `Eq`, `Hash`, and `Ord` are obvious candidates, yet they are unused in the current proposal:
+    -   Implementing `Eq`, `Hash`, or `Ord` for a collection does not require the handles to implement any of them.
+-   `Send` and `Sync` should definitely be kept out. `Allocator`-based stores could not use `NonNull<u8>` otherwise.
```

Comment:
Counter-example: a thread-local handle.

For compression reason, I can design a lightweight `Store` adapter which takes arbitrary-sized handles (typically `usize` or `NonNull<u8>`), stashes them into a thread-local, and hands out `u8` (or `[u8; 2/3/...]`) as a handle instead.

This naturally limits the number of handles, though not the total amount of memory allocated, and may be useful to represent a strongly connected graph for example.

Contrived? Yes. I blame my lack of imagination.

My point remains, though, that I see no reason to require `Send`/`Sync` and favor leaving it to the implementer of `Store`.

Reply:
Well then you better not be using uncovered `NonNull` as a handle, because you definitely want `Box<T, Global>` to be `Send`/`Sync` despite holding a handle!

Not requiring `Send`/`Sync` to be able to use thread-local handles with a non-thread-local store is a reasonable justification that I'm happy to go with, though even more niche than other applications. But it can't coexist with using `NonNull` handles, since you specifically *want* to cover the `!Send`/`!Sync` of pointers, which are only such as a practical default.